lib/filesystem.nix

```
@@ -306,106 +309,154 @@ in
+      anything in their scope, or in an ancestor of their scope.
+      This argument has no effect when `recurseIntoDirectory` is provided.
+
+    `directory`
+    : The directory to read package files from.
+
+    `recurseIntoDirectory`
+    : This argument is applied to the function which processes directories.
+    : Equivalently, this function takes `processDir` and `args`, and can modify arguments passed to `processDir`
+      (same as above) before calling it, as well as modify its output (which is then returned by `recurseIntoDirectory`).
+
+      :::{.note}
+      When `newScope` is set, the default `recurseIntoDirectory` is equivalent to:
+      ```nix
+      processDir: { newScope, ... }@args:
+        # create a new scope and mark it `recurseForDerivations`
+        lib.recurseIntoAttrs (lib.makeScope newScope (self:
+          # generate the attrset representing the directory, using the new scope's `callPackage` and `newScope`
+          processDir (args // {
+            inherit (self) callPackage newScope;
+          })
+        ))
+      ```
+      :::
+
+    `recurseArgs`
+    : Optional argument, which can be used to hold data used by `recurseIntoDirectory`
+
     # Examples
     :::{.example}
-    ## `lib.filesystem.packagesFromDirectoryRecursive` usage example
+    ## Basic use of `lib.packagesFromDirectoryRecursive`
 
     ```nix
     packagesFromDirectoryRecursive {
       inherit (pkgs) callPackage;
       directory = ./my-packages;
     }
     => { ... }
+    ```
+
+    In this case, `callPackage` will only search `pkgs` for a file's input parameters.
+    In other words, a file cannot refer to another file in the directory in its input parameters.
+    :::
 
-    lib.makeScope pkgs.newScope (
-      self: packagesFromDirectoryRecursive {
-        callPackage = self.callPackage;
-        directory = ./my-packages;
-      }
-    )
+    ::::{.example}
+    ## Create a scope for the nix files found in a directory
+    ```nix
+    packagesFromDirectoryRecursive {
+      inherit (pkgs) callPackage newScope;
+      directory = ./my-packages;
+    }
     => { ... }
     ```
 
+    For example, take the following directory structure:
+    ```
+    my-packages
+    ├── a.nix    → { b }: assert b ? b1; ...
+    └── b
+       ├── b1.nix  → { a }: ...
+       └── b2.nix
+    ```
+
+    Here, `b1.nix` can specify `{ a }` as a parameter, which `callPackage` will resolve as expected.
+    Likewise, `a.nix` receive an attrset corresponding to the contents of the `b` directory.
+
+    :::{.note}
+    `a.nix` cannot directly take as inputs packages defined in a child directory, such as `b1`.
     :::
+    ::::
   */
   packagesFromDirectoryRecursive =
+    let
+      inherit (lib.path) append;
+
+      # generate an attrset corresponding to a given directory
+      # this function is outside `packagesFromDirectoryRecursive`'s lambda expression,
+      #  to prevent accidentally using its parameters
```

```suggestion
      # Generate an attrset corresponding to a given directory.
      # This function is outside `packagesFromDirectoryRecursive`'s lambda expression
      # to prevent accidentally using its parameters.
```

Non-blocking slight reformatting. (I got this style preference from https://nedbatchelder.com/blog/201401/comments_should_be_sentences.html)